sway-core/src/semantic_analysis/cei_pattern_analysis.rs

```diff
@@ -0,0 +1,258 @@
+        match analysis_state {
+            CEIAnalysisState::LookingForInteraction => {
+                if effects_of_codeblock_entry(ast_node).contains(&Effect::Interaction) {
+                    analysis_state = CEIAnalysisState::LookingForStorageWrite;
+                    interaction_span = ast_node.span.clone();
+                }
+            }
+            CEIAnalysisState::LookingForStorageWrite => {
+                if effects_of_codeblock_entry(ast_node).contains(&Effect::StorageWrite) {
+                    warnings.push(CompileWarning {
+                        // TODO: or just use ast_node.span ? (i.e. warn on the storage ast node)
+                        span: Span::join(interaction_span, ast_node.span.clone()),
+                        // TODO: more info for this warning?
+                        warning_content: Warning::StorageWriteAfterInteraction,
+                    });
+                    return warnings;
+                }
+            }
+        }
+    }
+    warnings
+}
+
+fn effects_of_codeblock_entry(ast_node: &ty::TyAstNode) -> HashSet<Effect> {
+    match &ast_node.content {
+        ty::TyAstNodeContent::Declaration(decl) => effects_of_codeblock_decl(decl),
+        ty::TyAstNodeContent::Expression(expr)
+        | ty::TyAstNodeContent::ImplicitReturnExpression(expr) => effects_of_expression(expr),
+        ty::TyAstNodeContent::SideEffect => HashSet::new(),
+    }
+}
+
+fn effects_of_codeblock_decl(decl: &ty::TyDeclaration) -> HashSet<Effect> {
+    use crate::ty::TyDeclaration::*;
+    match decl {
+        VariableDeclaration(var_decl) => effects_of_expression(&(*var_decl).body),
+        // Declarations (except variable declarations) are not allowed in the body of a function
+        _ => HashSet::new(),
+    }
+}
+
+fn effects_of_expression(expr: &ty::TyExpression) -> HashSet<Effect> {
+    use crate::ty::TyExpressionVariant::*;
+    match &expr.expression {
+        Literal(_)
+        | VariableExpression{..}
+        | FunctionParameter
+        | Break
+        | Continue
+        // this type of assignment only mutates local variables and not storage
+        | Reassignment(_)
+        | AbiName(_)
+        | StorageAccess(_) => HashSet::new(),
+        StorageReassignment(storage_reassign) => {
+            let mut effs = HashSet::from([Effect::StorageWrite]);
+            effs.extend(effects_of_expression(&storage_reassign.rhs));
+            effs
+        },
+        LazyOperator {lhs, rhs, .. }
+        | ArrayIndex { prefix: lhs, index: rhs } => {
+            let mut effs = effects_of_expression(lhs);
+            let rhs_effs = effects_of_expression(rhs);
+            effs.extend(rhs_effs);
+            effs
+        },
+        Tuple { fields: exprs }
+        | Array { contents: exprs } => {
+            effects_of_expressions(exprs)
+        },
+        StructExpression {fields, ..} => {
+            effects_of_struct_expressions(fields)
+        },
+        CodeBlock(codeblock) => {
```

Comment:
Given this, I'm assuming that if I have something like
```rust
call_a_contract();
if condition {
    write_to_state()
}
```
then I will get a warning? This is good.

Reply:
Yes, you'll get a warning:

```
warning
  --> /Users/anton/fuel/sway/test/src/e2e_vm_tests/test_programs/should_pass/static_analysis/cei_pattern_violation_in_if_statement/src/main.sw:16:5
   |
14 |
15 |       // interaction
16 |       other_contract.deposit(amount);
   |  _____-
17 | |
18 | |     if amount == 42 {
19 | |         // effect -- therefore violation of CEI where effect should go before interaction
20 | |         let storage_key = 0x3dba0a4455b598b7655a7fb430883d96c9527ef275b49739e7b0ad12f8280eae;
21 | |         store(storage_key, ());
22 | |     }
   | |_____- Storage modification after external contract interaction in function or method "deposit". Consider making all storage writes before calling another contract
23 |     }
24 |   }
   |
____
```
The warning could be more precise actually and point at the method call / storage field assignment / asm instruction inside the `if` codeblock. This could be either improved in this PR or as a separate PR.